### ISSUE TITLE
Window MinGW unable to build due to pthread mingw and ldd

### ISSUE THREAD

**Issue by tilkinsc** [llvm/llvm-project#61539]
I am trying to rebuild llvm clang and family to use ldd built with mingw64. This means that clang++ depends on the mingw64 linker. Apparently pthread is needed to build llvm and clang and family. However, llvm doesn't add the -pthread when ninja in generated with cmake so it can. I was able to ctrl + f "  FLAGS = " and add -pthread to build successfully.

In short, explicit -pthread is needed when ldd depends on mingw64.

```bat
git clone --depth 1 --config core.autocrlf=false https://github.com/llvm/llvm-project
pushd llvm-project
cmake -DLLVM_ENABLE_PROJECTS="clang;lld;clang-tools-extra;lldb;cross-project-tests" -DCMAKE_INSTALL_PREFIX="C:/llvm/" -DLLVM_USE_LINKER=ldd -DCMAKE_BUILD_TYPE=Release -D -B build -S llvm -G "Ninja"
pushd build
ninja -j 24
popd
popd
```

**Comment by tilkinsc** [comment]
I was able to force the build to work by adding this to the CMakeLists.txt
```
// line 12
add_link_options(-pthread -femulated-tls)
add_compile_options(-pthread -femulated-tls)
```

I did a ninja install to my build dir to overwrite it and attempted to build again without t hese additives. It appears that it continues to use pthread.
```
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE
```

And my build failed with the same errors.

**Comment by zero9178** [comment]
A few things to note here that should help you:
* The best way to add compiler flags when using cmake is to add `-DCMAKE_CXX_FLAGS=...` to your command line. This adds the flags when compiling C++ source files and linking C++ code into a shared library or executable. Similarly, `-DCMAKE_C_FLAGS=...` for C flags (although I don't think that should be necessary here). So adding `-DCMAKE_CXX_FLAGS="-pthread -femulated-tls"` to your command line should do the trick here instead of modifying the `CMakeLists.txt`.
* LLVM does not require `pthreads` but you seem to be using libstdc++ on Windows which does. Why clang doesn't add this flag, I have no clue. 
* From your description it sounds like you want the resulting clang to use `lld` by default? What `-DLLVM_USE_LINKER=ldd` does (ignoring that you misspelt `lld` as `ldd`) is tell LLVM what linker to use while building LLVM, but not what the resulting clang binary should use. What you want is to add `-DCLANG_DEFAULT_LINKER=lld` to your cmake command. Also possibly `-DLLD_DEFAULT_LD_LLD_IS_MINGW=ON` for your convenience.

As a final note I'd also like to tell you that you do not have to build clang with clang if your current clang installation doesn't work out. You could also bootstrap clang with GCC or MSVC e.g. (though in the later case you'll likely want to add `-DLLVM_DEFAULT_TARGET_TRIPLE=x86_64-w64-windows-gnu` to your cmake command. By default the clang build takes the target triple from the compiler it was built from. This option overrides that. The `windows-gnu` triples are LLVM triples for MinGW basically). 

Hope this helps.

**Comment by tilkinsc** [comment]
For your first bullet point, I wasn't sure if that worked.

For your second point, I am not sure why I am depending on pthread in general when there are winthreads. I think the dependence of std::thread is whats doing it because mingw uses pthread in that instance otherwise no thread support. Syms in libstdc++ just wouldn't need to include pthread if there wasn't thread dependence.

For your third bullet point, I will give it a go. Ninja promised faster build speeds but with my ryzen 9 3900x (12 core 24 thread) it takes a while haha. I am sure I spelt it right, but I would also wager it would be good to reject invalid option passed to these arguments which it doesn't seem to do. I seen a few people in just a single day in discord who were passing like x86 and amd64 into a command.

I am bootstrapping with mingw which is gcc flavored to work with windows. I typically stay away from msvc to avoid being linked with it for all my projects. I don't agree with the state of standard libraries, in that in C its lacking in modernity and naming and never moves anywhere, and that in C++ its over-engineered to a point where everything is a unique pointer and static casted to everything else not to mention getting a full terminal page for 1 typo with a template, and gcc seems to be the the best case scenario out of those. Both of these languages shake in their boots compared to a language like C# who has an extensively robust and concise standard library and would definitely be using it for my compiler at any stage. I am currently researching llvm's clang to make use of IR to stay away from post fix type languages (name : type) and implement a compilable language in my own narcissistic vision (as programmers dream of doing one day). In which, I eventually want to interface with IR as the main route of code generation for the language itself (the dream of making something self compilable).

Because of my requirements listed here its undoubtedly burdensome when I have to compile with pthread itself, yet my program itself doesn't use pthread. It should already be linked in to the IR subsystem itself. This was why I was trying to bootstrap and move away from mingw.

I'll report my results here, and thank you for your input! If you have anything extra to say please feel welcome. :)

**Comment by zero9178** [comment]
> I think the dependence of std::thread is whats doing it because mingw uses pthread in that instance otherwise no thread support. Syms in libstdc++ just wouldn't need to include pthread if there wasn't thread dependence.

Correct. You can actually disable any kind of multi threading in llvm by using `-DLLVM_ENABLE_THREADS=OFF`. If you're lucky, it'll also drop the dependency on pthreads. 

> I am currently researching llvm's clang to make use of IR to stay away from post fix type languages (name : type) and implement a compilable language in my own narcissistic vision (as programmers dream of doing one day). In which, I eventually want to interface with IR as the main route of code generation for the language itself (the dream of making something self compilable).

Note that for doing this you don't need technically need to compile and use Clang. You'd just want to use LLVM as a library for creating a compiler. That is what Clang does as well. Clang is really just a C and C++ frontend that lowers to LLVM and then lets LLVM handle the rest. That is how languages using LLVM generally do it.

> Because of my requirements listed here its undoubtedly burdensome when I have to compile with pthread itself, yet my program itself doesn't use pthread. It should already be linked in to the IR subsystem itself. This was why I was trying to bootstrap and move away from mingw.

I am not 100% sure I fully understand what you mean here, but just so you know, what libraries the compiler requires and links against (pthreads e.g.), has no implication on what the output program produced by that compiler requires. So just because you need pthreads to compile LLVM, doesn't mean LLVM produced code needs it. 

As a final note, are you aware of distributions such as https://github.com/mstorsjo/llvm-mingw ? This is basically a whole MinGW repo using MinGW as libc and everything else from LLVM (so clang, lld, lldb, libc++, libunwind, sanitizers etc.). It also does not have any dependency on pthreads as libc++ uses Win32 threading directly. 
You could also replicate and build such a distro yourself from just the monorepo and MinGW (its what I do), but the repo should be of great help to you. I believe a simimlar setup is also available under MSYS2 using the `CLANG64` target or so.
Note that this is only a "clang toolchain". Not a LLVM library distribution that you'd use to create your own language

**Comment by HenriqueLourenconi** [comment]
Hello tilkinsc, were you able to compile with MinGW?

**Comment by tilkinsc** [comment]
I dont even remember doing this project. :)
